Wrappers/Python/cil/optimisation/functions/SubsetSumFunction.py

```diff
@@ -303,39 +297,366 @@ def memory_reset(self):
+
+    def __init__(self, functions, precond=None, update_frequency = 2, **kwargs):
+
+        super(SVRGFunction, self).__init__(functions)
 
         self.gradients_allocated = False
+    
         self.precond = precond
+        self.update_frequency = update_frequency
 
-        super(SGDFunction, self).__init__(functions)
+        # initialise the internal iteration counter. Used to check when to update the full gradient
+        self.iter = 0
 
     def gradient(self, x, out=None):
         """
-        Returns a vanilla stochastic gradient estimate, defined below.
-        For f = \sum f_i, the output is computed as follows:
+        Returns a variance-reduced approximate gradient, defined below.
+        For f = 1/num_subsets \sum_{i=1}^num_subsets f_i, the output is computed as follows:
             - choose a subset j with function next_subset()
-            - compute
-                subset_grad
-                with
-                - subset_grad is the gradient of function number j at current point
+            - check if full_grad_at_snapshot and snapshot should be updated
+            - if they shouldn't then compute
+                subset_gradient - subset_gradient_at_snapshot + full_grad_at_snapshot
+                where 
+                - subset_grad is the gradient of function number j at current_iterate
+                - subset_grad_at_snapshot is the gradient of function number j at snapshot in memory
+                    computed as full_gradient = 1/num_subsets \sum_{i=1}^num_subsets subset_gradient_{i} (snapshot)
+            - otherwise update full_grad_at_snapshot and snapshot
+            - this gives an unbiased estimator of the gradient
+        
+        Combined with the gradient step, the algorithm is guaranteed 
+        to converge if the functions f_i are convex and the step-size 
+        gamma satisfies
+            gamma <= 1/(4 * max L_i (P + 2))
+        where the gradient of each f_i is L_i - Lipschitz, and P is the number of inner iterations, that is, update_frequency * num_sebsets
 
-        where the gradient of each f_i is L_i - Lipschitz.
+        Reference:
+        Johnson, Rie; Zhang, Tong. 
+        "Accelerating stochastic gradient descent using predictive variance reduction." 
+        Advances in neural information processing systems. 2013.
         """
 
-        output = x * 0.0
+        if not self.gradients_allocated:
+            self.memory_init(x)         
+            
+        # Select the next subset (uniformly at random by default). This generates self.subset_num
         self.next_subset()
-        self.functions[self.subset_num].gradient(x, out=output)
 
-        return output
+        # In the first iteration and if the snapshot image has just been updated, the current iterate and snapshot will be the same
+        # thus tmp2 = 0
+        if self.iter == 0 or self.iter % (self.update_frequency * self.num_subsets) == 0:
+            self.tmp2 = 0.0 * self.full_gradient
```

Comment:
Should this if call `memory_update`? Maybe I am not understanding the method...

Reply:
Currently, `x` is updated before the `memory_update`. This is inefficiant. If the last algorithm iteration recomputes `memory_update` then the information is never used. This may not be a problem for small data but in large datasets a needless additional epoch is a lot of work.

Reply:
The if here doesn't meant to say "recompute the full gradient" but rather "if you just recomputed the full gradient, then this difference will be 0"

Reply:
Sort of (unless I'm misunderstanding the flow), you can either update the full gradient before the first iteration of a given epoch (assuming that update_frequency is 1), or after the last iteration of the previous epoch. In this code the latter is done. In both cases the current x and the snapshot will be the same (after the memory update).